opal/mca/hwloc/base/hwloc_base_util.c

```diff
@@ -2058,8 +2306,47 @@ char* opal_hwloc_base_get_topo_signature(hwloc_topology_t topo)
     endian = "unknown";
 #endif
 
-    asprintf(&sig, "%dN:%dS:%dL3:%dL2:%dL1:%dC:%dH:%s:%s",
-             nnuma, nsocket, nl3, nl2, nl1, ncore, nhwt, arch, endian);
+// The signature used to just contain number of nodes, sockets, cores,
+// hardware threads, etc.  With heterogeneous cgroups it's not hard to
+// have the counts come out the same even though the topologies of the
+// available elements are different enough to matter
+//
+// Example hardware:
+//    [..../..../..../....][..../..../..../....]
+//     0    4    8    12    16   20   24   28
+// cgset -r cpuset.cpus=10,11,14,15 mycgroup1
+// cgset -r cpuset.cpus=26,27,30,31 mycgroup2
+// the above cgroups would leave only these hardware threads active:
+//     mycgroup1: [~~~~/~~~~/~~HH/~~HH][~~~~/~~~~/~~~~/~~~~]
+//     mycgroup2: [~~~~/~~~~/~~~~/~~~~][~~~~/~~~~/~~HH/~~HH]
+// The non-WHOLE-SYSTEM topology would only contain the availabe part
+// of the tree and each host would report as 1 socket, 2 cores, etc
+//
+// I think a compressed list of the os_indexes in the available mask
+// should be a modest-sized addition to the string that would suffice
+// to make the signatures differ for different cgroups
+
+    int npu;
+    int *avail_pus;
+    char *avail_pus_string;
+    obj = NULL;
+    npu = hwloc_get_nbobjs_by_type(topo, HWLOC_OBJ_PU);
+    avail_pus = malloc(npu * sizeof(int));
+    if (!avail_pus) { return strdup("malloc failed"); }
+    npu = 0;
+    while ((obj = hwloc_get_next_obj_by_type(topo, HWLOC_OBJ_PU, obj)) != NULL) {
+        avail_pus[npu++] = obj->os_index;
+    }
```

Comment:
You're basically duplicating the topology entire cpuset into your own array of indexes here. The hwloc bitmap likely has everything needed to replace this. For instance, you may iterate over PU physical indexes with:
```
hwloc_bitmap_foreach_begin(i, hwloc_topology_get_topology_cpuset(topo))
{ /* use physical index i */
} hwloc_bitmap_foreach_end();
```
and there are functions to find the first, last, next bits, the number of bits, etc.

Reply:
Yeah, we had a discussion on a telecon about this - we don't want to intro a new regex method into the system, especially a release branch. We just want to use the hwloc print methods and attach the result to the existing signature.

Reply:
Wouldn't I need the os_index for each "i" though?  So either I'd have to construct those into a mask or do a lookup on each logical "i" for its os_index.  That seems quite a bit heavier than "a[i]" to access the desired data.

Plus the hwloc-awareness would have to extend all the way into both "mycompress" functions which are currently written generically for an integer array.